File: src/api/sdks/javascript/playergold-sdk.js
/**
 * PlayerGold SDK para JavaScript
 * SDK para integración de blockchain PlayerGold en juegos web
 * @version 1.0.0
 */

class PlayerGoldSDK {
    /**
     * Constructor del SDK
     * @param {Object} config - Configuración del SDK
     * @param {string} config.apiKey - API key del desarrollador
     * @param {string} config.apiUrl - URL base de la API (opcional)
     */
    constructor(config = {}) {
        this.apiKey = config.apiKey;
        this.apiUrl = config.apiUrl || 'http://localhost:5000/api/v1';
        this.authToken = null;
        this.tokenExpiration = null;
    }

    /**
     * Inicializa el SDK y autentica con la API
     * @returns {Promise<boolean>} - True si la autenticación fue exitosa
     */
    async initialize() {
        try {
            await this.authenticate();
            console.log('PlayerGold SDK initialized successfully');
            return true;
        } catch (error) {
            console.error('Failed to initialize PlayerGold SDK:', error);
            return false;
        }
    }

    /**
     * Autentica con la API y obtiene token JWT
     * @private
     * @returns {Promise<void>}
     */
    async authenticate() {
        const response = await fetch(`${this.apiUrl}/auth/token`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                api_key: this.apiKey
            })
        });

        if (!response.ok) {
            throw new Error(`Authentication failed: ${response.statusText}`);
        }

        const data = await response.json();
        this.authToken = data.token;
        this.tokenExpiration = Date.now() + (data.expires_in * 1000);
    }

    /**
     * Asegura que el token esté válido
     * @private
     * @returns {Promise<void>}
     */
    async ensureAuthenticated() {
        // Re-autenticar si el token está por expirar (5 minutos antes)
        if (!this.authToken || Date.now() >= (this.tokenExpiration - 300000)) {
            await this.authenticate();
        }
    }

    /**
     * Realiza una petición HTTP autenticada
     * @private
     * @param {string} endpoint - Endpoint de la API
     * @param {Object} options - Opciones de fetch
     * @returns {Promise<Object>} - Respuesta JSON
     */
    async request(endpoint, options = {}) {
        await this.ensureAuthenticated();

        const headers = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.authToken}`,
            ...options.headers
        };

        const response = await fetch(`${this.apiUrl}${endpoint}`, {
            ...options,
            headers
        });

        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.error || response.statusText);
        }

        return response.json();
    }

    /**
     * Obtiene el estado de la red
     * @returns {Promise<Object>} - Estado de la red
     */
    async getNetworkStatus() {
        const response = await fetch(`${this.apiUrl}/network/status`);
        
        if (!response.ok) {
            throw new Error(`Failed to get network status: ${response.statusText}`);
        }
        
        return response.json();
    }

    /**
     * Obtiene el saldo de una dirección
     * @param {string} address - Dirección de la wallet
     * @returns {Promise<number>} - Saldo en $PRGLD
     */
    async getBalance(address) {
        const data = await this.request(`/balance/${address}`);
        return data.balance;
    }

    /**
     * Crea una nueva transacción
     * @param {Object} transaction - Datos de la transacción
     * @param {string} transaction.fromAddress - Dirección origen
     * @param {string} transaction.toAddress - Dirección destino
     * @param {number} transaction.amount - Cantidad a transferir
     * @param {string} transaction.privateKey - Clave privada para firmar
     * @param {number} [transaction.fee=0.01] - Fee de la transacción
     * @returns {Promise<string>} - Hash de la transacción
     */
    async createTransaction({ fromAddress, toAddress, amount, privateKey, fee = 0.01 }) {
        const data = await this.request('/transaction', {
            method: 'POST',
            body: JSON.stringify({
                from_address: fromAddress,
                to_address: toAddress,
                amount,
                private_key: privateKey,
                fee
            })
        });

        return data.transaction_hash;
    }

    /**
     * Obtiene información de una transacción
     * @param {string} txHash - Hash de la transacción
     * @returns {Promise<Object>} - Información de la transacción
     */
    async getTransaction(txHash) {
        return this.request(`/transaction/${txHash}`);
    }

    /**
     * Obtiene el historial de transacciones de una dirección
     * @param {string} address - Dirección de la wallet
     * @param {Object} options - Opciones de paginación
     * @param {number} [options.page=1] - Número de página
     * @param {number} [options.perPage=20] - Transacciones por página
     * @returns {Promise<Object>} - Historial de transacciones
     */
    async getTransactionHistory(address, { page = 1, perPage = 20 } = {}) {
        return this.request(`/transactions/history/${address}?page=${page}&per_page=${perPage}`);
    }

    /**
     * Obtiene información de un bloque
     * @param {number} blockIndex - Índice del bloque
     * @returns {Promise<Object>} - Información del bloque
     */
    async getBlock(blockIndex) {
        return this.request(`/block/${blockIndex}`);
    }

    /**
     * Espera la confirmación de una transacción
     * @param {string} txHash - Hash de la transacción
     * @param {number} [requiredConfirmations=3] - Confirmaciones requeridas
     * @param {number} [timeout=60000] - Timeout en milisegundos
     * @returns {Promise<Object>} - Información de la transacción confirmada
     */
    async waitForConfirmation(txHash, requiredConfirmations = 3, timeout = 60000) {
        const startTime = Date.now();
        const pollInterval = 2000; // 2 segundos

        while (Date.now() - startTime < timeout) {
            try {
                const txInfo = await this.getTransaction(txHash);

                if (txInfo.status === 'confirmed' && txInfo.confirmations >= requiredConfirmations) {
                    return txInfo;
                }

                await new Promise(resolve => setTimeout(resolve, pollInterval));
            } catch (error) {
                console.warn('Error checking transaction status:', error);
                await new Promise(resolve => setTimeout(resolve, pollInterval));
            }
        }

        throw new Error('Transaction confirmation timeout');
    }
}

// Clase auxiliar para gestión de wallets en el navegador
class PlayerGoldWallet {
    /**
     * Genera una nueva wallet
     * @returns {Object} - Wallet con address y privateKey
     */
    static generateWallet() {
        // En producción, usar una librería criptográfica apropiada
        console.warn('This is a demo implementation. Use proper cryptography in production.');
        
        const randomBytes = new Uint8Array(32);
        crypto.getRandomValues(randomBytes);
        
        const privateKey = Array.from(randomBytes)
            .map(b => b.toString(16).padStart(2, '0'))
            .join('');
        
        // Generar address desde private key (simplificado)
        const address = 'PRGLD' + privateKey.substring(0, 40);
        
        return {
            address,
            privateKey
        };
    }

    /**
     * Guarda una wallet en localStorage
     * @param {Object} wallet - Wallet a guardar
     * @param {string} password - Password para encriptar
     */
    static saveWallet(wallet, password) {
        // En producción, encriptar apropiadamente
        const encrypted = btoa(JSON.stringify(wallet));
        localStorage.setItem('playergold_wallet', encrypted);
    }

    /**
     * Carga una wallet desde localStorage
     * @param {string} password - Password para desencriptar
     * @returns {Object|null} - Wallet o null si no existe
     */
    static loadWallet(password) {
        const encrypted = localStorage.getItem('playergold_wallet');
        
        if (!encrypted) {
            return null;
        }

        try {
            return JSON.parse(atob(encrypted));
        } catch (error) {
            console.error('Failed to load wallet:', error);
            return null;
        }
    }
}

// Exportar para uso en Node.js y navegador
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { PlayerGoldSDK, PlayerGoldWallet };
}

// Exportar para uso con ES6 modules
if (typeof window !== 'undefined') {
    window.PlayerGoldSDK = PlayerGoldSDK;
    window.PlayerGoldWallet = PlayerGoldWallet;
}

File: src/api/sdks/unity/PlayerGoldSDK.cs
/*
 * PlayerGold SDK para Unity
 * SDK para integración de blockchain PlayerGold en juegos Unity
 */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;

namespace PlayerGold.SDK
{
    /// <summary>
    /// Cliente principal del SDK de PlayerGold para Unity
    /// </summary>
    public class PlayerGoldSDK : MonoBehaviour
    {
        [Header("API Configuration")]
        [SerializeField] private string apiUrl = "http://localhost:5000/api/v1";
        [SerializeField] private string apiKey;
        
        private string authToken;
        private float tokenExpirationTime;
        
        // Singleton
        private static PlayerGoldSDK instance;
        public static PlayerGoldSDK Instance
        {
            get
            {
                if (instance == null)
                {
                    GameObject go = new GameObject("PlayerGoldSDK");
                    instance = go.AddComponent<PlayerGoldSDK>();
                    DontDestroyOnLoad(go);
                }
                return instance;
            }
        }
        
        private void Awake()
        {
            if (instance != null && instance != this)
            {
                Destroy(gameObject);
                return;
            }
            instance = this;
            DontDestroyOnLoad(gameObject);
        }
        
        /// <summary>
        /// Inicializa el SDK con la API key
        /// </summary>
        public void Initialize(string apiKey, string apiUrl = null)
        {
            this.apiKey = apiKey;
            if (!string.IsNullOrEmpty(apiUrl))
            {
                this.apiUrl = apiUrl;
            }
            
            StartCoroutine(AuthenticateCoroutine());
        }
        
        /// <summary>
        /// Autentica con la API y obtiene token JWT
        /// </summary>
        private IEnumerator AuthenticateCoroutine()
        {
            string url = $"{apiUrl}/auth/token";
            
            var requestData = new Dictionary<string, string>
            {
                { "api_key", apiKey }
            };
            
            string jsonData = JsonUtility.ToJson(new ApiKeyRequest { api_key = apiKey });
            
            using (UnityWebRequest request = UnityWebRequest.Post(url, jsonData, "application/json"))
            {
                yield return request.SendWebRequest();
                
                if (request.result == UnityWebRequest.Result.Success)
                {
                    TokenResponse response = JsonUtility.FromJson<TokenResponse>(request.downloadHandler.text);
                    authToken = response.token;
                    tokenExpirationTime = Time.time + response.expires_in;
                    
                    Debug.Log("PlayerGold SDK: Autenticación exitosa");
                }
                else
                {
                    Debug.LogError($"PlayerGold SDK: Error de autenticación - {request.error}");
                }
            }
        }
        
        /// <summary>
        /// Obtiene el saldo de una dirección
        /// </summary>
        public void GetBalance(string address, Action<float> onSuccess, Action<string> onError)
        {
            StartCoroutine(GetBalanceCoroutine(address, onSuccess, onError));
        }
        
        private IEnumerator GetBalanceCoroutine(string address, Action<float> onSuccess, Action<string> onError)
        {
            yield return EnsureAuthenticated();
            
            string url = $"{apiUrl}/balance/{address}";
            
            using (UnityWebRequest request = UnityWebRequest.Get(url))
            {
                request.SetRequestHeader("Authorization", $"Bearer {authToken}");
                
                yield return request.SendWebRequest();
                
                if (request.result == UnityWebRequest.Result.Success)
                {
                    BalanceResponse response = JsonUtility.FromJson<BalanceResponse>(request.downloadHandler.text);
                    onSuccess?.Invoke(response.balance);
                }
                else
                {
                    onError?.Invoke(request.error);
                }
            }
        }
        
        /// <summary>
        /// Crea una transacción
        /// </summary>
        public void CreateTransaction(
            string fromAddress,
            string toAddress,
            float amount,
            string privateKey,
            Action<string> onSuccess,
            Action<string> onError
        )
        {
            StartCoroutine(CreateTransactionCoroutine(fromAddress, toAddress, amount, privateKey, onSuccess, onError));
        }
        
        private IEnumerator CreateTransactionCoroutine(
            string fromAddress,
            string toAddress,
            float amount,
            string privateKey,
            Action<string> onSuccess,
            Action<string> onError
        )
        {
            yield return EnsureAuthenticated();
            
            string url = $"{apiUrl}/transaction";
            
            TransactionRequest txRequest = new TransactionRequest
            {
                from_address = fromAddress,
                to_address = toAddress,
                amount = amount,
                private_key = privateKey,
                fee = 0.01f
            };
            
            string jsonData = JsonUtility.ToJson(txRequest);
            
            using (UnityWebRequest request = UnityWebRequest.Post(url, jsonData, "application/json"))
            {
                request.SetRequestHeader("Authorization", $"Bearer {authToken}");
                
                yield return request.SendWebRequest();
                
                if (request.result == UnityWebRequest.Result.Success)
                {
                    TransactionResponse response = JsonUtility.FromJson<TransactionResponse>(request.downloadHandler.text);
                    onSuccess?.Invoke(response.transaction_hash);
                }
                else
                {
                    onError?.Invoke(request.error);
                }
            }
        }
        
        /// <summary>
        /// Obtiene el estado de una transacción
        /// </summary>
        public void GetTransaction(string txHash, Action<TransactionInfo> onSuccess, Action<string> onError)
        {
            StartCoroutine(GetTransactionCoroutine(txHash, onSuccess, onError));
        }
        
        private IEnumerator GetTransactionCoroutine(string txHash, Action<TransactionInfo> onSuccess, Action<string> onError)
        {
            yield return EnsureAuthenticated();
            
            string url = $"{apiUrl}/transaction/{txHash}";
            
            using (UnityWebRequest request = UnityWebRequest.Get(url))
            {
                request.SetRequestHeader("Authorization", $"Bearer {authToken}");
                
                yield return request.SendWebRequest();
                
                if (request.result == UnityWebRequest.Result.Success)
                {
                    TransactionInfo response = JsonUtility.FromJson<TransactionInfo>(request.downloadHandler.text);
                    onSuccess?.Invoke(response);
                }
                else
                {
                    onError?.Invoke(request.error);
                }
            }
        }
        
        /// <summary>
        /// Obtiene el estado de la red
        /// </summary>
        public void GetNetworkStatus(Action<NetworkStatus> onSuccess, Action<string> onError)
        {
            StartCoroutine(GetNetworkStatusCoroutine(onSuccess, onError));
        }
        
        private IEnumerator GetNetworkStatusCoroutine(Action<NetworkStatus> onSuccess, Action<string> onError)
        {
            string url = $"{apiUrl}/network/status";
            
            using (UnityWebRequest request = UnityWebRequest.Get(url))
            {
                yield return request.SendWebRequest();
                
                if (request.result == UnityWebRequest.Result.Success)
                {
                    NetworkStatus response = JsonUtility.FromJson<NetworkStatus>(request.downloadHandler.text);
                    onSuccess?.Invoke(response);
                }
                else
                {
                    onError?.Invoke(request.error);
                }
            }
        }
        
        /// <summary>
        /// Asegura que el token esté válido
        /// </summary>
        private IEnumerator EnsureAuthenticated()
        {
            if (string.IsNullOrEmpty(authToken) || Time.time >= tokenExpirationTime - 300)
            {
                yield return AuthenticateCoroutine();
            }
        }
    }
    
    // Clases de datos
    [Serializable]
    public class ApiKeyRequest
    {
        public string api_key;
    }
    
    [Serializable]
    public class TokenResponse
    {
        public string token;
        public int expires_in;
    }
    
    [Serializable]
    public class BalanceResponse
    {
        public string address;
        public float balance;
        public string timestamp;
    }
    
    [Serializable]
    public class TransactionRequest
    {
        public string from_address;
        public string to_address;
        public float amount;
        public string private_key;
        public float fee;
    }
    
    [Serializable]
    public class TransactionResponse
    {
        public string transaction_hash;
        public string status;
        public string message;
    }
    
    [Serializable]
    public class TransactionInfo
    {
        public string status;
        public int confirmations;
        public TransactionData transaction;
    }
    
    [Serializable]
    public class TransactionData
    {
        public string from_address;
        public string to_address;
        public float amount;
        public float fee;
        public float timestamp;
    }
    
    [Serializable]
    public class NetworkStatus
    {
        public int chain_length;
        public int last_block_index;
        public string last_block_hash;
        public int pending_transactions;
        public int difficulty;
    }
}
